File: 03-Vectorization/Homework/code/problem_01/no_avx.c
#include<stdio.h>
#include<stdlib.h>
#include<time.h>

#define TEST_ITERATIONS 10
#define PROCESS_ITERATIONS 100
#define ARRAY_SIZE 6400000

void add(int *a, int *b);
double run_test();

int main() {
    printf("============ SCALAR RUNTIME ===========\n");
    for(int i = 1; i <= TEST_ITERATIONS; i++) {
        printf("TEST %d: %f s\n", i, run_test());
    }
    return 0;
}

void add(int *a, int *b) {
    for(int i = 0; i < ARRAY_SIZE; i++) {
        a[i] += b[i];
    }
}

double run_test() {
    // Allocated memory for array a and b.
    int *a = (int*)malloc(ARRAY_SIZE * sizeof(int));
    int *b = (int*)malloc(ARRAY_SIZE * sizeof(int));

    // Initialize value inside each array.
    for(int i = 0; i < ARRAY_SIZE; i++) {
        a[i] = i;
        b[i] = i;
    }

    // Run an experiment
    clock_t begin_time = clock();
    for(int i = 0; i < PROCESS_ITERATIONS; i++) {
        add(a, b);
    }
    clock_t end_time = clock();

    // Free allocated memory
    free(a);
    free(b);

    // Return executed time
    double run_time = ((double) end_time - begin_time) / CLOCKS_PER_SEC;
    return run_time;
}

File: 03-Vectorization/Homework/code/problem_02/no_avx.py
import time

TEST_ITERATIONS = 10
ARRAY_SIZE = 6400000


def main():
    print("============ SCALAR RUNTIME ===========")
    for i in range(TEST_ITERATIONS):
        print(f"TEST {i + 1}: {run_test():.3f} s")


def run_test():
    # Create two integer arrays
    a = list(range(ARRAY_SIZE))
    b = list(range(ARRAY_SIZE))

    # Run an experiment
    begin_time = time.time()
    for i in range(ARRAY_SIZE):
        a[i] += b[i]
    end_time = time.time()

    # Return executed time
    return end_time - begin_time


if __name__ == "__main__":
    main()
